### PR TITLE
Adding overlay helpers and implementing initial overlay methods - 🐙

### PR DESCRIPTION
**Pull request by TheMcMurder**

### CODE REVIEW

File: src/overlays.helpers.js [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+			.slice(2, -5)
+	, 10) & 0XFFFFFF).toString(16).toUpperCase()
+	result = result.split('').concat([0,0,0,0,0,0]).slice(0,6).join('')
+	const rgba = [`0x${result.slice(0, 2)}`, `0x${result.slice(2, 4)}`, `0x${result.slice(4, 6)}`]
+	return `rgba(${parseInt(rgba[0])}, ${parseInt(rgba[1])}, ${parseInt(rgba[2])}, ${opacity})`
+}
+
+export function createOverlayWithText (opts, elementToAppendChild) {
+	const div = elementToAppendChild.appendChild(document.createElement('div'))
+	div.style.width = opts.overlay.width || '100%'
```
Comment (TheMcMurder) [inline src/overlays.helpers.js]: overrides. This is my initial pass at overrides. I doubt anyone will need them, but I think this is a valuable first pass on the settings

**Comment by TheMcMurder** [comment]
@joeldenning I wasn't sure if this would merge cleanly into master. It seems that it will. We don't necessarily have to do it right now. I just wanted a brief review of the refactor

File: src/overlays.helpers.js [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+		parseInt(string, 36)
+			.toExponential()
+			.slice(2, -5)
+	, 10) & 0XFFFFFF).toString(16).toUpperCase()
+	result = result.split('').concat([0,0,0,0,0,0]).slice(0,6).join('')
+	const rgba = [`0x${result.slice(0, 2)}`, `0x${result.slice(2, 4)}`, `0x${result.slice(4, 6)}`]
+	return `rgba(${parseInt(rgba[0])}, ${parseInt(rgba[1])}, ${parseInt(rgba[2])}, ${opacity})`
+}
+
+export function createOverlayWithText (opts, elementToAppendChild) {
```
Comment (joeldenning) [inline src/overlays.helpers.js]: Like we talked about, maybe short circuit if the element doesn't exist.

File: src/single-spa-canopy.js [hunk 3]
```diff
@@ -90,8 +96,21 @@ function bootstrap(opts) {
...
 			const el = getDomEl(opts);
+			el.style.position = 'relative'
+			if (opts.overlay.selectors) {
+				overlayArray = overlayArray.concat(
+					opts.overlay.selectors.map((selector, i) => {
+						return createOverlayWithText(opts, document.querySelector(selector))
+					})
+				)
+			} else {
+				overlayArray = overlayArray.concat(createOverlayWithText(opts, el))
```
Comment (joeldenning) [inline src/single-spa-canopy.js]: Should we do this no matter what, even if selectors or provided?

Or maybe provide a configuration option `overlay.overlayRootElement` that defaults to true?

File: src/overlays.helpers.js [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+	return div
+}
+
+export function addOverlayEventListener () {
+	window.addEventListener('keypress', function (evt) {
+		if (evt.code === 'KeyD' && evt.shiftKey && evt.ctrlKey) {
+			localStorage.setItem('cp:single-spa:overlay', !JSON.parse(localStorage.getItem('cp:single-spa:overlay')))
+			window.dispatchEvent(new CustomEvent('cp:show-overlay-keypress'))
+		} else if (evt.keyCode === 83 && evt.shiftKey && evt.ctrlKey) {
```
Comment (joeldenning) [inline src/overlays.helpers.js]: Should we just get rid of the temporary mode for showing things? Does this line of code work?

File: src/single-spa-canopy.js [hunk 5]
```diff
@@ -1,5 +1,10 @@
 import {initializeHotReloading} from './hot-reload.js';
 import deepMerge from 'deepmerge';
+import { shouldShowOverlay, getColorFromString, createOverlayWithText, addOverlayEventListener } from './overlays.helpers.js'
+
+if (!window._overlayListenerDefined) {
```
Comment (joeldenning) [inline src/single-spa-canopy.js]: Couldn't this be done inside of `overlays.helpers.js`?

**Pull request closed without merging**